### DISCUSSION TITLE
Orderless completion doesn’t work in comments?

### DISCUSSION THREAD

**Discussion by kings2u** [company-mode/company-mode#1211] [Q&A] [answered]
I use the Orderless package for `completion-styles`, and the author recommends at https://github.com/oantolin/orderless/#company to use `(setq orderless-component-separator "[ &]")` in order to sort Company results with Orderless instead of Company. This works great while typing code in an Emacs-Lisp buffer using the built-in Company backends, but if I’m trying to get completion when typing in a code _comment_, using `&` as a separator between search strings no longer works. This is the case despite setting `company-dabbrev-code-everywhere` to `t`, though I’m not sure that’s relevant. 

Does anyone have advice on how to use Orderless for searching Company in all possible contexts?

**Comment by dgutov** [comment]
Unfortunately, Orderless only works through `completion-styles`, and when we're in a comment, that usually means using the `company-dabbrev` backend, which is a classic Company backend which doesn't work through CAPF API.

You can take a look https://github.com/PythonNut/company-flx, but it provides a different matching method than Orderless.

**Reply by kings2u** [reply]
Thank you for the quick reply and the suggestion! There is also the user option to change `company-search-regexp-function` to `company-search-words-in-any-order-regexp` that basically accomplishes what Orderless does, but I’ve noticed it is slightly inconsistent. For example, when writing code in an elisp buffer, if you type `tex`, then execute `company-filter-candidates` and type `boo`, you get an error even though `bookmark-edit-annotation-text-function` and `bookmark-read-annotation-text-function` are still candidates that contain both the `tex` and `boo` strings. I’ve found this error occurs when the second searched string is the first word in the candidate (as in the example I gave). Can I ask, is this a bug or the intended behavior?

**Reply by dgutov** [reply]
Good find. Does it work better with the change below?

```diff
diff --git a/company.el b/company.el
index 1fffe96..2e36d52 100644
--- a/company.el
+++ b/company.el
@@ -2256,7 +2256,7 @@ each one wraps a part of the input string."
   (let ((re (funcall company-search-regexp-function text))
         (i 0))
     (cl-dolist (line lines)
-      (when (string-match-p re line (length company-prefix))
+      (when (string-match-p re line)
         (cl-return i))
       (cl-incf i))))
 
```

**Reply by kings2u** [reply]
Indeed it does! The only thing it doesn’t fix is that the search string `bo` in `bookmark-edit-annotation-text-function` and `bookmark-read-annotation-text-function` is not highlighted with the `company-tooltip-search` face like the other candidates containing `tex` and `bo` are. Not a huge deal, but it’s something to note if there’s an easy fix for that too. I look forward to this package update :)

**Reply by dgutov** [reply]
There was a similar problem there, should be fixed now too. :+1: 

And it was actually a leftover from the old design where we only did prefix matching -- so the code made sure to skip the prefix when doing the search, assuming the user would want to only search in the remaining part of the completions (since they begin the same). This assumption no longer holds, so it's easier to remove it than to do some kind of adaptation.

**Reply by dgutov** [reply]
If you like how it works now, you can write your own reply and mark is as the answer ;)

**Reply by kings2u** [reply]
Everything works swimmingly now, thank you again :) I will of course write a summary and mark it as the answer.

**Comment by kings2u** [comment, marked as answer]
The above reply from @dgutov solved the issue for me. For those who stumble onto this issue, I recommend changing the user option `company-search-regexp-function` to `company-search-words-in-any-order-regexp`, then commence a search using `company-filter-candidates` (bound to <kbd>C-s</kbd>). This accomplishes the most common use case for the Orderless package of narrowing based on out-of-order strings.
